plugins/dynamic-discovery-backend/src/plugin.ts

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  coreServices,
+  createBackendPlugin,
+} from '@backstage/backend-plugin-api';
+import { createRouter } from './service/router';
+import { GatewayDiscoveryService } from './implementations/discovery/GatewayDiscoveryService';
+import { LeafNodeDiscoveryService } from './implementations/discovery/LeafNodeDiscoveryService';
+
+/**
+ * dynamicDiscoveryPlugin backend plugin
+ *
+ * @public
+ */
+export default createBackendPlugin({
+  pluginId: 'dynamic-discovery',
+  register(env) {
+    env.registerInit({
+      deps: {
+        httpRouter: coreServices.httpRouter,
+        logger: coreServices.logger,
+        discovery: coreServices.discovery,
+      },
+      async init({ httpRouter, logger, discovery }) {
+        console.log(discovery);
+        if (
+          !(discovery instanceof GatewayDiscoveryService) &&
+          !(discovery instanceof LeafNodeDiscoveryService)
+        ) {
+          throw new Error(
+            'Invalid discovery service, you must install the dynamicDiscoveryServiceFactory as well.',
+          );
+        }
+        if (discovery instanceof GatewayDiscoveryService) {
```

Comment:
Would be good if this could be done in a different way that doesn't break the abstraction

Reply:
updated in [aa51900](https://github.com/backstage/backstage/pull/24539/commits/aa519003aafdf3eee4d38f6f756c8ccd14f742db)